compiler/rustc_errors/src/json.rs

```diff
@@ -180,6 +183,38 @@ struct Diagnostic {
     children: Vec<Diagnostic>,
     /// The message as rustc would render it.
     rendered: Option<String>,
+    /// Extra tool metadata
+    tool_metadata: ToolMetadata,
+}
+
+macro_rules! encode_fields {
+    ($enc:expr, $s:expr, $idx:expr, [ $($name:ident),+$(,)? ]) => {
+        {
+            let mut idx = $idx;
+            $(
```

Comment:
Per [this point](https://github.com/rust-lang/rust/pull/72603/files#r559808807), maybe we can change the desugaring to

```rust
let $struct_name { $($name),+ } = $s;
let mut idx = 0;
$(
    $enc.emit_struct_field(
        stringify!($name),
        idx,
        |enc| $name.encode(enc),
    )?;
    idx += 1;
)+
idx
```

Reply:
Hm, that works, but since the whole point is that we're leaving out some fields to special-case them it will also need a list of those so they can be matched.